scripts/prepare-localise.ts

```diff
@@ -0,0 +1,143 @@
+    cwd,
+    externals: [
+      ...Object.keys({ ...pkg.dependencies, ...pkg.devDependencies, ...pkg.peerDependencies }),
+      ...builtins,
+    ],
+    input: inputs[0],
+    optimized: flags.includes('--optimized'),
+    watch: flags.includes('--watch'),
+  };
+
+  if (flags.includes('--optimized')) {
+    const { stdout } = await command(`yarn info ${pkg.name} -R --json`);
+    const {
+      children: { Dependencies },
+    } = JSON.parse(stdout);
+
+    const list = (Dependencies as { descriptor: string; locator: string }[])
+      .filter(({ descriptor }) => !descriptor.startsWith('@types/'))
+      .filter(({ locator }) => !locator.includes('workspace:'))
+      .map(({ locator, descriptor }) => {
+        const version = locator.split('npm:')[1];
+        const name = descriptor.split(/@npm|@virtual/)[0];
+        return `${name}@${version}`;
+      });
+
+    const location = join(
+      __dirname,
+      '../../storybook-temp-modules',
+      cwd.replace(join(__dirname, '..'), '')
+    );
+
+    await ensureDir(location);
+    await remove(join(location, 'node_modules'));
+
+    const afterInit = await command(`yarn init -i=classic -y`, { cwd: location });
+    if (afterInit.failed) {
+      console.error(afterInit.stderr);
+      throw new Error('Failed to init package');
+    }
+    const afterAdd = await command(`npx add-dependencies ./package.json ${list.join(' ')}`, {
+      cwd: location,
+    });
+    if (afterAdd.failed) {
+      console.error(afterAdd.stderr);
+      throw new Error('Failed to add dependencies');
+    }
+    const afterInstall = await command(
+      `yarn install --ignore-scripts --ignore-engines --ignore-optional --no-bin-links`,
+      { cwd: location }
+    );
+    if (afterInstall.failed) {
+      console.error(afterInstall.stderr);
+      throw new Error('Failed to install dependencies');
+    }
+
+    await rm('-rf', join(location, 'node_modules', '@types'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.md'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.markdown'));
+    await rm('-rf', join(location, 'node_modules', '**', 'bower.json'));
+    await rm('-rf', join(location, 'node_modules', '**', 'component.json'));
+    await rm('-rf', join(location, 'node_modules', '**', 'jest.config.js'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.test.*'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.png'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.jpg'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.jpeg'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.gif'));
+    await rm('-rf', join(location, 'node_modules', '**', '.*'));
+    await rm('-rf', join(location, 'node_modules', '**', '*.d.ts'));
+
+    await command('find . -type d -empty -print -delete', { cwd: location });
+
+    await cp('-R', join(location, 'node_modules'), join(cwd, 'dist'));
+    await rm('-rf', location);
+  }
```

Comment:
This is where the magic happens!

- I ask `yarn` what the dependencies for the current package are, their exact version according to the lockfile, this is important
- I init a new yarn package in a sibling directory of the monorepo root project
- I I add all the listen dependencies to this package.json
- I run `yarn install` there, with some flags to keep the noise to a minimum
- I run a bunch of cleanup commands, in a hope to minimise the amount useless files
- I copy the `node_modules` directory into `dist` of my package in the monorepo.

The above combined with using `require.resolve` should mean that the localised dependencies should get priority in the node resolver.